File: benchmarks/compute/fft/Go/fft.go
package main

import (
	"fmt"
	"math"
	"math/cmplx"
	"math/rand"
	"os"
	"strconv"
	"time"
)

const (
	Small    = 0
	Standard = 1
	Large    = 2
)

// 编译时可覆盖此值：go build -ldflags="-X main.Dataset=Large"
var Dataset = Standard

var (
	M, N int
)

func init() {
	switch Dataset {
	case Small:
		M, N = 64, 64
	case Large:
		M, N = 2048, 2048
	default:
		M, N = 512, 512
	}
}

func initArray(data [][]float64, seed int64) {
	rand.Seed(seed)
	for i := range data {
		for j := range data[i] {
			data[i][j] = rand.Float64()
		}
	}
}

func fft(n int, in []float64, out []complex128) {
	if n == 1 {
		out[0] = complex(in[0], 0)
		return
	}

	half := n / 2
	evenIn, oddIn := make([]float64, half), make([]float64, half)
	evenOut, oddOut := make([]complex128, half), make([]complex128, half)

	for i := 0; i < half; i++ {
		evenIn[i] = in[2*i]
		oddIn[i] = in[2*i+1]
	}

	fft(half, evenIn, evenOut)
	fft(half, oddIn, oddOut)

	for k := 0; k < half; k++ {
		t := cmplx.Exp(complex(0, -2*math.Pi*float64(k)/float64(n))) * oddOut[k]
		out[k] = evenOut[k] + t
		out[k+half] = evenOut[k] - t
	}
}

func computeFFT(data [][]float64, result [][]complex128) {
	for i := range data {
		fft(N, data[i], result[i])
	}
}

func printMatrix(result [][]complex128) {
	rows := int(math.Min(8, float64(len(result))))
	cols := int(math.Min(8, float64(len(result[0]))))

	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			fmt.Printf("(%.2f,%.2f) ", real(result[i][j]), imag(result[i][j]))
		}
		fmt.Println()
	}
}

func main() {
	// 初始化二维数组
	data := make([][]float64, M)
	result := make([][]complex128, M)
	for i := range data {
		data[i] = make([]float64, N)
		result[i] = make([]complex128, N)
	}

	// 参数处理（默认种子42）
	seed := int64(42)
	if len(os.Args) > 1 {
		if s, err := strconv.ParseInt(os.Args[1], 10, 64); err == nil {
			seed = s
		}
	}

	initArray(data, seed)

	start := time.Now()
	computeFFT(data, result)
	elapsed := time.Since(start)

	fmt.Printf("Finished FFT in %.3f seconds\n", elapsed.Seconds())
	printMatrix(result)
}

File: benchmarks/compute/fft/Python/fft.py
import sys
import time
import math
import cmath
import random

# 数据集设定
SMALL, STANDARD, LARGE = 0, 1, 2
DATASET = STANDARD  # 可根据需要改为 SMALL 或 LARGE

if DATASET == SMALL:
    M, N = 64, 64
elif DATASET == LARGE:
    M, N = 2048, 2048
else:
    M, N = 512, 512

def init_array(m, n, seed):
    random.seed(seed)
    data = [[random.random() for _ in range(n)] for _ in range(m)]
    return data

def fft(n, real_in):
    if n == 1:
        return [complex(real_in[0], 0.0)]
    half = n // 2
    even = fft(half, real_in[0::2])
    odd = fft(half, real_in[1::2])
    result = [0] * n
    for k in range(half):
        twiddle = cmath.exp(-2j * math.pi * k / n) * odd[k]
        result[k] = even[k] + twiddle
        result[k + half] = even[k] - twiddle
    return result

def compute_fft(m, n, data):
    result = []
    for i in range(m):
        result.append(fft(n, data[i]))
    return result

def print_matrix(m, n, result):
    for i in range(min(m, 8)):
        for j in range(min(n, 8)):
            val = result[i][j]
            print(f"({val.real:.2f},{val.imag:.2f})", end=" ")
        print()

def main():
    seed = int(sys.argv[1]) if len(sys.argv) > 1 else 42
    data = init_array(M, N, seed)
    
    start = time.time()
    result = compute_fft(M, N, data)
    end = time.time()

    print(f"Finished FFT in {end - start:.3f} seconds")
    print_matrix(M, N, result)

if __name__ == "__main__":
    main()

